### PR TITLE
Add maximum depth and complexity instrumentation

### PR DESCRIPTION
**Pull request by andimarek**
This PR adds several things: 
- a query visitor and based on that a query reducer (NOT PUBLIC API)
- AbortExecutionException which lets you write Instrumentations to abort the current execution
- MaxQueryDepthInstrumentation and MaxQueryComplexityInstrumentation:
This  Instrumentations are build on top of the new query reducer and lets you prevent queries which exceeds a certain query depth or query complexity.

I decided to build it on n top of Instrumentation because it worked. :-) I didn't want to add a new abstraction/layer if not needed.

The query visitor/reducer is not public api because I don't think it is needed at the moment and I didn't feel comfortable making it public.

### CODE REVIEW

**Comment by andimarek** [comment]
I am not completely sure about the Exception: maybe we can handle it better, like returning an error or so.

File: src/main/java/graphql/schema/GraphQLObjectType.java [hunk 1]
```diff
@@ -72,8 +72,7 @@ public GraphQLFieldDefinition getFieldDefinition(String name) {
 
     /**
      * @return This returns GraphQLInterface or GraphQLTypeReference instances, if the type
-     * references are not resolved yet. After they are resolved it contains only GraphQLInterface.
-     * Reference resolving happens when a full schema is built.
+     * references are not resolved yet. After they are resolved it contains only GraphQLInterface. * Reference resolving happens when a full schema is built.
      */
```
Comment (bbakerman) [inline src/main/java/graphql/schema/GraphQLObjectType.java]: JavaDoc weirdness here I think

File: src/test/groovy/graphql/GraphQLTest.groovy [hunk 2]
```diff
@@ -579,4 +581,40 @@ class GraphQLTest extends Specification {
...
+
+        def graphql = GraphQL.newGraphQL(schema).instrumentation(maximumQueryDepthInstrumentation).build()
+
+        when:
+        def result = graphql.execute('{ field {field {field {field {scalar}}}} }')
+
+        then:
+        result.errors.size() == 1
+        result.errors[0].message.contains("maximum query depth exceeded")
+
```
Comment (bbakerman) [inline src/test/groovy/graphql/GraphQLTest.groovy]: Maybe use `where:` data driven tests and input a series of different depths.... just to be sure

File: src/main/java/graphql/schema/GraphQLObjectType.java (hunk 1 above)
Comment (andimarek) [reply src/main/java/graphql/schema/GraphQLObjectType.java]: yes ... don't know how this happened, now fixed

File: src/main/java/graphql/analysis/MaxQueryComplexityInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            int totalComplexity = valuesByParent.get(null).stream().mapToInt(Integer::intValue).sum();
+            if (totalComplexity > maxComplexity) {
+                throw new AbortExecutionException("maximum query complexity exceeded " + totalComplexity + " > " + maxComplexity);
+            }
+        };
+    }
+
+    private Integer calculateComplexity(QueryVisitorEnvironment environment, int childCount) {
+        // interface call here ...
+        return 1 + childCount;
```
Comment (bbakerman) [inline src/main/java/graphql/analysis/MaxQueryComplexityInstrumentation.java]: I am guessing you are going to fix this up.  I think you should

File: src/main/java/graphql/execution/AbortExecutionException.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import graphql.ErrorType;
+import graphql.GraphQLError;
+import graphql.GraphQLException;
+import graphql.PublicApi;
+import graphql.language.SourceLocation;
+
+import java.util.List;
+
+@PublicApi
+public class AbortExecutionException extends GraphQLException implements GraphQLError {
```
Comment (bbakerman) [inline src/main/java/graphql/execution/AbortExecutionException.java]: JavaDoc

**Comment by bbakerman** [comment]
Looking further into this  - this is a really neat PR - the traversal code WILL prove useful in the future and the combination of Instrumentation as a way to inject behaviour into a graphql execution is cool!

File: src/test/groovy/graphql/GraphQLTest.groovy (hunk 2 above)
Comment (andimarek) [reply src/test/groovy/graphql/GraphQLTest.groovy]: done

File: src/main/java/graphql/analysis/MaxQueryComplexityInstrumentation.java (hunk 3 above)
Comment (andimarek) [reply src/main/java/graphql/analysis/MaxQueryComplexityInstrumentation.java]: done

File: src/main/java/graphql/analysis/FieldComplexityCalculator.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package graphql.analysis;
+
+import graphql.PublicApi;
+
+@PublicApi
```
Comment (bbakerman) [inline src/main/java/graphql/analysis/FieldComplexityCalculator.java]: JavaDoc

File: src/main/java/graphql/analysis/MaxQueryComplexityInstrumentation.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import static graphql.Assert.assertNotNull;
+
+/**
+ * Prevents execution if the query complexity is greater than the specified maxDepth
+ */
+@PublicApi
```
Comment (bbakerman) [inline src/main/java/graphql/analysis/MaxQueryComplexityInstrumentation.java]: Javadoc is wrong here:  talks of max Depth

copy pasta?

**Comment by sbilello** [comment]
Would you use the same approach with instrumentation to grant or deny access to a particular field? 
How would you make fine-grained RBAC access at the field level? 
Can I take inspiration from this one or it should be something different?

**Comment by kaqqao** [comment]
@sbilello  It's certainly a possibility. See my answer [here](https://stackoverflow.com/questions/51764092/how-to-get-query-mutation-operation-name/51845557#51845557).

**Comment by sbilello** [comment]
@kaqqao thanks would you use a similar approach with `Instrumentation` to achieve a Graphql node object `rate-limit based on QPS` and maybe plug Redis to keep track of the graphql API usage in a distributed systems?
I am evaluating if I should build this a spring module and externalize a generic approach so the final developer can decide to wrap with its favorite rate limit library.
The goal is to be able to define the limit in terms of the QPS number of query per second against a particular graphql object as well as the number of nodes returned.

**Comment by Cpt76** [comment]
I'm using Spring GraphQL, but I can't find cost analysis support there...

**Comment by bclozel** [comment]
@Cpt76 there is a dedicated issue in Spring for GraphQL, for a higher level support: https://github.com/spring-projects/spring-graphql/issues/469

Note that this does not prevent you from using the GraphQL Java instrumentation right now. It's not clear from your comment what you're looking for.

**Comment by Cpt76** [comment]
> 

Thank you, understood. I will use GraphQL Java directly.

**Comment by bclozel** [comment]
@Cpt76 to be clear, you can use Spring for GraphQL and configure this GraphQL Java instrumentation in your app.

**Pull request merged**